### ISSUE TITLE
HLSL fragment shaders generated by glslangValidator fail on MSL.

### ISSUE THREAD

**Issue by repi** [KhronosGroup/SPIRV-Cross#245]
@billhollings

I have a fairly simple HLSL pixel shader that I've converted to SPIRV with glslang, and then trying to convert to Metal with SPIRV-Cross. 

The fragment shader has an input structure of attributes from the vertex shader, HLSL code:

```
struct VertexOutput
{
	float3 worldPosition  : POSITION;
	float3 worldNormal    : NORMAL;
	float2 texCoord       : TEXCOORD;
};

PixelOutput geometryPs(VertexOutput input)
{
    ...
}
```

And this looks correct when using SPIRV-Cross to generate GLSL:

```
struct VertexOutput
{
    vec3 worldPosition;
    vec3 worldNormal;
    vec2 texCoord;
};

layout(location = 0) in VertexOutput _input;

PixelOutput _geometryPs(VertexOutput _input_1)
{
    emissiveColor = texture(SPIRV_Cross_Combinedg_emissiveg_sampler, _input_1.texCoord).xyz;
}

void main()
{
    VertexOutput _input_1 = _input;
    VertexOutput param = _input_1;
    PixelOutput flattenTemp = _geometryPs(param);
```

But when using `SPIRV-Cross --msl` to generate Metal the passing of the vertex output structure from the main function to the other function is using the wrong types. It uses `VertexOutput` struct in the main function instead of the generated `geometryPs_in` struct, so it seems to be mismatching the real input struct and the generated one , both in the `geometryPs_out` and in the `_geometryPs` functions.

```
struct VertexOutput
{
    float3 worldPosition;
    float3 worldNormal;
    float2 texCoord;
};

struct geometryPs_in
{
    float3 VertexOutput_worldPosition [[user(locn0)]];
    float3 VertexOutput_worldNormal [[user(locn1)]];
    float2 VertexOutput_texCoord [[user(locn2)]];
};


PixelOutput _geometryPs(thread const VertexOutput& input, ...)
{
    emissiveColor = g_emissive.sample(g_sampler, in.VertexOutput_texCoord).xyz;
    ...
}

fragment geometryPs_out geometryPs(geometryPs_in in [[stage_in]], ...)
{
    geometryPs_out out = {};
    VertexOutput input = input;
    VertexOutput param = input;
    PixelOutput flattenTemp = _geometryPs(param, v_142, g_emissive, g_sampler, g_albedo, g_normal, g_roughness, g_metalness, g_ao);

```

Is this is a known issue with the experimental Metal support, a new bug or am I missing something? Or perhaps something that comes from using HLSL to SPIRV as a source?

99% of our shaders use this pattern, and we only use HLSL, so getting this supported would be most excellent.

Have attached the shaders here, includes the source HLSL, the generated SPIRV, Metal and GLSL shaders:
[GeometryPs_shaders.zip](https://github.com/KhronosGroup/SPIRV-Cross/files/1196719/GeometryPs_shaders.zip)

Cheers,
Johan

**Comment by HansKristian-ARM** [comment]
MSL flattens structs and I/O blocks into a single stage_in/stage_out structure. The output seems correct to me at a glance, but I'll let @billhollings have a closer look.

**Comment by repi** [comment]
The generated code in this example is definitely invalid Metal code which one can see with the 'in' parameter is referenced as 'input' instead in the geometryPs and in the _geometryPs it is the opposite case, the parameter is named 'input' but referred to as 'in' in the function body. So it is somehow mixing up the original struct for the vertex shader outputs ("VertexOutput") with the SPIRV-Cross generated one ("geometryPs_in")

**Comment by HansKristian-ARM** [comment]
Ah, right, now I see what you mean, that's clearly a bug.

**Comment by billhollings** [comment]
Fixed in PR #254.

I have included a test shader built from @HansKristian-ARM's GLSL above.

I have also tested it with the `GeometryPs_shader` provided above...and it builds and compiles with the Metal compiler fine. I have not included that shader as a test in this repo since it is a proprietary shader.

**Comment by repi** [comment]
Can confirm that the latest glslang + SPIRV-Cross fixes this issue, excellent!

Thanks!
